src/rebar_compiler_dag.erl

```diff
@@ -142,62 +226,167 @@ store_dag(G, File, CritMeta) ->
+            ?DEBUG("Source ~ts is gone, deleting previous ~ts file if it exists ~ts", [Source, Ext, Target]),
             file:delete(Target),
             digraph:del_vertex(G, Source),
             mark_dirty(G),
             true
     end.
 
-%% @private Return what should be the base name of an erl file, relocated to the
-%% target directory. For example:
-%% target_base("ebin/", "src/my_module.erl") -> "ebin/my_module"
-target_base(OutDir, Source) ->
-    filename:join(OutDir, filename:basename(Source, ".erl")).
-
-%% @private a file has been found to change or wasn't part of the DAG before,
-%% and must be added, along with all its dependencies.
-add_to_dag(G, Compiler, InDirs, Source, LastModified, SourceDir) ->
-    AbsIncls = Compiler:dependencies(Source, SourceDir, InDirs),
-    digraph:add_vertex(G, Source, LastModified),
-    digraph:del_edges(G, digraph:out_edges(G, Source)),
-    %% Deps must be explored before the module itself
-    [begin
-         update(G, Compiler, InDirs, [Incl]),
-         digraph:add_edge(G, Source, Incl)
-     end || Incl <- AbsIncls],
-    mark_dirty(G),
-    AbsIncls.
-
-%% @private change status propagation: if the dependencies of a file have
-%% been updated, mark the last_modified time for that file to be equivalent
-%% to its most-recently-changed dependency; that way, nested header file
-%% change stamps are propagated to the final module.
-%% This is required because at some point the module is compared to its
-%% associated .beam file's last-generation stamp to know if it requires
-%% rebuilding.
-%% The responsibility for this is however diffuse across various modules.
-update_max_modified_deps(G, Source) ->
-    MaxModified = lists:foldl(
-        fun(File, Acc) ->
+maybe_rm_vertex(G, Source) ->
+    case filelib:is_regular(Source) of
+        true ->
+            exists;
+        false ->
+            digraph:del_vertex(G, Source),
+            mark_dirty(G)
+    end.
+
+%% Add dependencies of a given file to the DAG. If the file is not found yet,
+%% mark its timestamp to 0, which means we have no info on it.
+%% Source files will be covered at a later point in their own scan, and
+%% non-source files are going to be covered by `populate_deps/3'.
+prepopulate_deps(G, Compiler, InDirs, Source, DepOpts, Status) ->
+    SourceDir = filename:dirname(Source),
+    AbsIncls = case erlang:function_exported(Compiler, dependencies, 4) of
+        false ->
+            Compiler:dependencies(Source, SourceDir, InDirs);
+        true ->
+            Compiler:dependencies(Source, SourceDir, InDirs, DepOpts)
+    end,
+    %% the file hasn't been visited yet; set it to existing, but with
+    %% a last modified value that's null so it gets updated to something new.
+    [digraph:add_vertex(G, Src, 0) || Src <- AbsIncls,
+                                      digraph:vertex(G, Src) =:= false],
+    %% drop edges from deps that aren't included!
+    [digraph:del_edge(G, Source, Path) || Status == old,
```

Comment:
digraph:del_edge/3 does not exist, is "Source" just extra here?

Reply:
Not sure, I think that was in my bus-written branch that got added through your change set, but I'll patch this up right away.

Reply:
This was probably never called because it was an additional change for files that stopped tracking other files and cleaning them up, which we likely don't have tests for!